module/zfs/metaslab.c

```diff
@@ -438,39 +449,214 @@ metaslab_class_destroy(metaslab_class_t *mc)
 
 	for (int i = 0; i < spa->spa_alloc_count; i++) {
 		metaslab_class_allocator_t *mca = &mc->mc_allocator[i];
+		avl_destroy(&mca->mca_tree);
+		mutex_destroy(&mca->mca_lock);
 		ASSERT(mca->mca_rotor == NULL);
-		zfs_refcount_destroy(&mca->mca_alloc_slots);
+		ASSERT0(mca->mca_reserved);
 	}
 	mutex_destroy(&mc->mc_lock);
 	multilist_destroy(&mc->mc_metaslab_txg_list);
 	kmem_free(mc, offsetof(metaslab_class_t,
 	    mc_allocator[spa->spa_alloc_count]));
 }
 
-int
+void
 metaslab_class_validate(metaslab_class_t *mc)
 {
-	metaslab_group_t *mg;
-	vdev_t *vd;
+#ifdef ZFS_DEBUG
+	spa_t *spa = mc->mc_spa;
+
+	/*
+	 * Must hold one of the spa_config locks.
+	 */
+	ASSERT(spa_config_held(spa, SCL_ALL, RW_READER) ||
+	    spa_config_held(spa, SCL_ALL, RW_WRITER));
+
+	for (int i = 0; i < spa->spa_alloc_count; i++) {
+		metaslab_class_allocator_t *mca = &mc->mc_allocator[i];
+		metaslab_group_t *mg, *rotor;
+
+		ASSERT0(avl_numnodes(&mca->mca_tree));
+		ASSERT0(mca->mca_reserved);
+
+		if ((mg = rotor = mca->mca_rotor) == NULL)
+			continue;
+		do {
+			metaslab_group_allocator_t *mga = &mg->mg_allocator[i];
+			vdev_t *vd = mg->mg_vd;
+
+			ASSERT3P(vd->vdev_top, ==, vd);
+			ASSERT(vd->vdev_mg == mg || vd->vdev_log_mg == mg);
+			ASSERT3P(mg->mg_class, ==, mc);
+			ASSERT3P(vd->vdev_ops, !=, &vdev_hole_ops);
+			ASSERT0(mga->mga_queue_depth);
+		} while ((mg = mg->mg_next) != rotor);
+	}
+#endif
+}
+
+void
+metaslab_class_balance(metaslab_class_t *mc, boolean_t onsync)
+{
+	metaslab_group_t *mg, *first;
 
 	/*
 	 * Must hold one of the spa_config locks.
 	 */
 	ASSERT(spa_config_held(mc->mc_spa, SCL_ALL, RW_READER) ||
 	    spa_config_held(mc->mc_spa, SCL_ALL, RW_WRITER));
 
-	if ((mg = mc->mc_allocator[0].mca_rotor) == NULL)
-		return (0);
+	if (onsync)
+		metaslab_class_validate(mc);
+
+	if (mc->mc_groups == 0) {
+		if (onsync)
+			mc->mc_alloc_throttle_enabled = B_FALSE;
+		mc->mc_alloc_max = UINT64_MAX;
+		return;
+	}
+
+	if (onsync) {
+		mc->mc_alloc_io_size = (3 * mc->mc_alloc_io_size +
+		    metaslab_aliquot / 8) / 4;
```

Comment:
This calculation has some hardcoded values.  Can you add a comment describing what it's doing?

Reply:
Yes.  Added a comment.